packages/grafana-ui/src/components/TimeSeries/utils.ts

```diff
@@ -262,6 +262,58 @@ export const preparePlotConfigBuilder: UPlotConfigPrepFn<{ sync: DashboardCursor
 
   builder.scaleKeys = [xScaleKey, yScaleKey];
 
+  // if hovered value is null, how far we may scan left/right to hover nearest non-null
+  const hoverProximityPx = 15;
+
+  let cursor: Partial<uPlot.Cursor> = {
+    // this scans left and right from cursor position to find nearest data index with value != null
+    // TODO: do we want to only scan past undefined values, but halt at explicit null values?
+    dataIdx: (self, seriesIdx, hoveredIdx, cursorXVal) => {
+      let seriesData = self.data[seriesIdx];
+
+      if (seriesData[hoveredIdx] == null) {
+        let nonNullLft = hoveredIdx,
+          nonNullRgt = hoveredIdx,
+          i;
+
+        i = hoveredIdx;
+        while (nonNullLft === hoveredIdx && i-- > 0) {
+          if (seriesData[i] != null) {
+            nonNullLft = i;
+          }
+        }
+
+        i = hoveredIdx;
+        while (nonNullRgt === hoveredIdx && i++ < seriesData.length) {
+          if (seriesData[i] != null) {
+            nonNullRgt = i;
+          }
+        }
+
+        let xVals = self.data[0];
+
+        let curPos = self.valToPos(cursorXVal, 'x');
+        let rgtPos = self.valToPos(xVals[nonNullRgt], 'x');
+        let lftPos = self.valToPos(xVals[nonNullLft], 'x');
+
+        let lftDelta = curPos - lftPos;
+        let rgtDelta = rgtPos - curPos;
+
+        if (lftDelta <= rgtDelta) {
+          if (lftDelta <= hoverProximityPx) {
+            hoveredIdx = nonNullLft;
+          }
+        } else {
+          if (rgtDelta <= hoverProximityPx) {
+            hoveredIdx = nonNullRgt;
+          }
+        }
+      }
```

Comment:
Awesome job on this! I know it could be tricky but would be nice if complex code like this with this many conditions could have some unit tests, here are the tests for the old graph tooltip find data hover point: https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/graph/specs/graph_tooltip.test.ts#L43

Reply:
i'd like to hold off on adding tests until we figure out what the new normal should be, i think the proximity thresholds and left/right scanning logic is likely to change here. also it may be a bit more challenging since we actually measure pixel distance to the found values and are not simply relying on nearest-non null, so the test scaffold may need to be more complex that the previous tests. i'll add a new issue to add tests, but want to land this in the next beta round to get some feedback.

Reply:
this logic was extracted into uPlot core recently (will be in next release).

https://github.com/leeoniya/uPlot/pull/918

on grafana side it will be converted to more simple uPlot config options and will be exposed as a panel option:

https://github.com/grafana/grafana/pull/81421